packages/component-base/src/slot-observer.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * @license
+ * Copyright (c) 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+
+/**
+ * A helper for observing slot changes.
+ */
+export class SlotObserver {
+  constructor(slot, callback) {
+    /** @type HTMLSlotElement */
+    this.slot = slot;
+
+    /** @type Function */
+    this.callback = callback;
+
+    /** @type {Node[]} */
+    this._storedNodes = [];
```

Comment:
**nit:** Could be a Set.

```suggestion
    this._storedNodes = new Set();
```

Reply:
Updated to use `forEach()` on this array, which requires array (in case of `Set` we [don't have keys](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach#description)).